pkg/distributor/influx.go

```diff
@@ -0,0 +1,194 @@
+	defer spanLogger.Span.Finish()
+
+	spanLogger.SetTag("content_type", r.Header.Get("Content-Type"))
+	spanLogger.SetTag("content_encoding", r.Header.Get("Content-Encoding"))
+	spanLogger.SetTag("content_length", r.ContentLength)
+
+	pts, bytesRead, err := influxpush.ParseInfluxLineReader(ctx, r, maxSize)
+	level.Debug(spanLogger).Log("msg", "decodeAndConvert complete", "bytesRead", bytesRead, "metric_count", len(pts), "err", err)
+	if err != nil {
+		level.Error(logger).Log("msg", "failed to parse Influx push request", "err", err)
+		return bytesRead, err
+	}
+
+	req.Timeseries = pts
+	return bytesRead, nil
+}
+
+// InfluxHandler is a http.Handler which accepts Influx Line protocol and converts it to WriteRequests.
+func InfluxHandler(
+	maxRecvMsgSize int,
+	requestBufferPool util.Pool,
+	sourceIPs *middleware.SourceIPExtractor,
+	retryCfg RetryConfig,
+	push PushFunc,
+	pushMetrics *PushMetrics,
+	logger log.Logger,
+) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		ctx := r.Context()
+		logger := utillog.WithContext(ctx, logger)
+		if sourceIPs != nil {
+			source := sourceIPs.Get(r)
+			if source != "" {
+				logger = utillog.WithSourceIPs(source, logger)
+			}
+		}
+
+		tenantID, err := tenant.TenantID(ctx)
+		if err != nil {
+			level.Warn(logger).Log("msg", "unable to obtain tenantID", "err", err)
+			return
+		}
+
+		pushMetrics.IncInfluxRequest(tenantID)
+
+		var bytesRead int
+
+		supplier := func() (*mimirpb.WriteRequest, func(), error) {
+			rb := util.NewRequestBuffers(requestBufferPool)
+			var req mimirpb.PreallocWriteRequest
+
+			if bytesRead, err = influxRequestParser(ctx, r, maxRecvMsgSize, rb, &req, logger); err != nil {
+				err = httpgrpc.Error(http.StatusBadRequest, err.Error())
+				rb.CleanUp()
+				return nil, nil, err
+			}
+
+			cleanup := func() {
+				mimirpb.ReuseSlice(req.Timeseries)
+				rb.CleanUp()
+			}
+			return &req.WriteRequest, cleanup, nil
+		}
+
+		pushMetrics.ObserveInfluxUncompressedBodySize(tenantID, float64(bytesRead))
+
+		req := newRequest(supplier)
+		// https://docs.influxdata.com/influxdb/cloud/api/v2/#tag/Response-codes
+		if err := push(ctx, req); err != nil {
+			if errors.Is(err, context.Canceled) {
+				level.Warn(logger).Log("msg", "push request canceled", "err", err)
+				w.WriteHeader(statusClientClosedRequest)
+				return
+			}
+			if errors.Is(err, influxio.ErrReadLimitExceeded) {
+				// TODO(alexg): One thing we have seen in the past is that telegraf clients send a batch of data
+				// if it is too big they should respond to the 413 below, but if a client doesn't understand this
+				// it just sends the next batch that is even bigger. In the past this has had to be dealt with by
+				// adding rate limits to drop the payloads.
+				level.Warn(logger).Log("msg", "request too large", "err", err, "bytesRead", bytesRead, "maxMsgSize", maxRecvMsgSize)
```

Comment:
Is this really a TODO? What is pending to be done here?

Reply:
Removed the TODO. This is more an operational issue, something that I plan on taking a look at during the dev/ops testing and will add it to the epic.

When the `telegraf` sends a payload that is too big it just tries it again a bit later with even more data, and so the problem never seems to fix itself.

We need to documented method/runbook for dealing with this situation (which is easy to trigger given we can control the maximum payload size we accept) otherwise it never resolves itself. We need to test how the `telegraf` agent responds to different return codes, or come up with another plan.

For now the code does the right thing and logs the error and rejects the payload.